### PR TITLE
feat: add radio component

### PR DESCRIPTION
**Pull request by josokinas**
## Purpose

Add Radio component.

## Approach

Implements Radio component.

This closes issue #8.

## Testing

On Storybook preview.

## Risks

N/A

### CODE REVIEW

**Comment by josokinas** [comment]
Focus state is not yet implemented, as it is different from our current one used with the mixin.

File: packages/core/src/components/radio/radio.scss [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    &.-disabled {
+      border-color: helpers.color('border-disabled');
+    }
+
+    > :not([aria-hidden='true']) {
+      display: flex;
+      flex-direction: column;
+      padding-left: helpers.space(1.625);
+      padding-right: helpers.space(0.375);
+      white-space: pre-line;
```
Comment (rabelloo) [inline packages/core/src/components/radio/radio.scss:41]: 🤔  What's the intention here?

File: packages/core/src/components/radio/radio.scss [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    &.-disabled {
+      border-color: helpers.color('border-disabled');
+    }
+
+    > :not([aria-hidden='true']) {
+      display: flex;
+      flex-direction: column;
+      padding-left: helpers.space(1.625);
+      padding-right: helpers.space(0.375);
```
Comment (rabelloo) [inline packages/core/src/components/radio/radio.scss:40]: The paddings can be substituted by `gap` on the parent:
```scss
.ods-radio-label {
  gap: helpers.space(1.5);
}
```

The column flex is also not necessary, you can simply render blocks, e.g.
```html
<div>Label</div>
<div>Description</div>
```

File: packages/core/src/components/radio/radio.scss [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+      }
+    }
+
+    &.-invalid {
+      border-color: helpers.color('border-negative');
+    }
+
+    &.-invalid + span {
+      background-color: helpers.color('content-negative');
+    }
```
Comment (rabelloo) [inline packages/core/src/components/radio/radio.scss:94]: 💅🏻  [nitpick]

```suggestion

      + span {
        background-color: helpers.color('content-negative');
      }
    }
```

File: packages/core/src/components/radio/radio.scss [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  .ods-radio-input {
+    @include mixins.interactive();
+
+    appearance: none;
+    border: 2px solid helpers.color('border-separator');
+    margin: 0;
+    transition: border-color 0.15s ease-out;
```
Comment (rabelloo) [inline packages/core/src/components/radio/radio.scss:65]: Since `.-disabled` changes the background color, should probably transition that too.

```suggestion
    transition: 0.15s ease-out;
    transition-property: background-color, border-color;
```

**Review by rabelloo — CHANGES_REQUESTED** [review]

**Comment by rabelloo** [comment]
I've also found a very subtle UI issue.
Because the input is 22px of height, but the text has a 24px line-height, it's not correctly centered.

There are two possible fixes, need input from @Subcide probably.

| Current | 24px input | 22px line-height |
| --- | --- | --- |
| Screenshot 2021-01-06 at 15 28 51 | Screenshot 2021-01-06 at 15 30 17 | Screenshot 2021-01-06 at 15 30 34 |

**Comment by rabelloo** [comment]
Oh there's also no way to distinguish disabled radios checked/unchecked right now.
May I suggest something like this? @Subcide 

image

File: packages/core/src/components/radio/radio.scss [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+@mixin Radio() {
+  $boundary: helpers.space(6);
+  $size: helpers.space(2.75);
+  $border-width: 1px;
+
+  .ods-radio-label {
+    @include mixins.interactive();
+
+    $padding: ($boundary - $size) / 2;
```
[the remark this answers is not in the document]
Comment (josokinas) [reply packages/core/src/components/radio/radio.scss]: This is only specific to `.ods-radio-label` - putting into variable as it needs adjustment for "bordered" variation. Other variable `$input-size` is used in both `.ods-radio-label` and `.ods-radio-input` classes.

File: packages/react/src/components/radio/radio.react.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  const [checked, setChecked] = useState(false);
+
+  useEffect(() => {
+    const timeout = setTimeout(() => {
+      if (checked) {
+        setChecked(false);
+      }
+    }, 3000);
+    return () => clearTimeout(timeout);
+  }, [checked]);
```
[the remark this answers is not in the document]
Comment (josokinas) [reply packages/react/src/components/radio/radio.react.stories.tsx:15]: I was thinking about two, however issue is with `storyOf` where once selected you aren't able to unselect.

Obviously, main story could have two and other stories could have such (with timeout).

Unless you have another suggestion?

File: packages/core/src/components/radio/radio.scss (hunk 1 above)
Comment (josokinas) [reply packages/core/src/components/radio/radio.scss:40]: I suspect description can be long text with new lines? @Subcide would it be a case?

**Comment by josokinas** [comment]
> There are two possible fixes...

I think @Subcide mentioned something about input size, maybe 24px is too big. Might need to adjust line-height then, which here is the default of a font.

> Oh there's also no way to distinguish disabled radios checked/unchecked right now.

Yes, this is true - designs do not show such case, let's wait for the input.

File: packages/core/src/components/radio/radio.scss (hunk 3 above)
Comment (josokinas) [reply packages/core/src/components/radio/radio.scss:96]: I can't do such because of our linting 🙈

Screenshot 2021-01-06 at 17 59 55

I'm pretty sure even now I'm  actually just overcoming a problem...

File: packages/core/src/components/radio/radio.scss (hunk 5 above)
[the remark this answers is not in the document]
Comment (rabelloo) [reply packages/core/src/components/radio/radio.scss]: Wait, what? I don't see `$input-size` anywhere.

```ts
  $boundary: helpers.space(6); // used only in .ods-radio-label
  $size: helpers.space(2.75); // shared between .ods-radio-label and .ods-radio-input
  $border-width: 1px; // used only in .ods-radio-label
```

File: packages/core/src/components/radio/radio.scss (hunk 2 above)
Comment (josokinas) [reply packages/core/src/components/radio/radio.scss:39]: Thanks, I can use `gap` instead of left padding for sure, however value should be `1.625` because it should be `13px` if I understand correctly:

Screenshot 2021-01-06 at 18 03 32

File: packages/react/src/components/radio/radio.react.stories.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (rabelloo) [reply packages/react/src/components/radio/radio.react.stories.tsx:15]: Other stories have multiple radios too, all we need is a static `name` and we can switch between the selected one.

There's always refreshing the page for resetting state anyways.

File: packages/core/src/components/radio/radio.scss (hunk 5 above)
[the remark this answers is not in the document]
Comment (josokinas) [reply packages/core/src/components/radio/radio.scss]: You're looking into outdated version, now it's:

Screenshot 2021-01-06 at 18 05 56

File: packages/core/src/components/radio/radio.scss (hunk 1 above)
Comment (rabelloo) [reply packages/core/src/components/radio/radio.scss:40]: Ah I see, a downside of accepting only strings.

If we do JSX we can have `<br />` tags in there.

```tsx
<Radio>
  Label
  <Description>
    Line 1<br />
    Line 2
  </Description>
</Radio>
```

File: packages/core/src/components/radio/radio.scss (hunk 3 above)
Comment (rabelloo) [reply packages/core/src/components/radio/radio.scss:96]: Ah yes. It works for `:checked` and `:disabled` because they are pseudos.
No worries then.

File: packages/core/src/components/radio/radio.scss (hunk 2 above)
Comment (rabelloo) [reply packages/core/src/components/radio/radio.scss:39]: Yup, you're right. Just used `1.5` for simplicity.
And I guess the right padding can be applied to `label > span` instead of all its children.

It's a bit unintuitive how our spacings in there work, but it's neat that we don't need event listeners and can use the `label`, I like it.

File: packages/core/src/components/radio/radio.scss (hunk 5 above)
[the remark this answers is not in the document]
Comment (rabelloo) [reply packages/core/src/components/radio/radio.scss]: Ah ok, looks good then.

**Comment by josokinas** [comment]
> I've also found a very subtle UI issue.

This now has been fixed.

> Oh there's also no way to distinguish disabled radios checked/unchecked right now.

And this now has been added.

**Comment by rabelloo** [comment]
Checked, disabled and bordered, I don't suppose the background should be blue, right?

image

**Comment by josokinas** [comment]
> Checked, disabled and bordered, I don't suppose the background should be blue, right?

True, issue with specificity. I did a "fix" for now, which is not nice. Will take a look at better solution.

File: packages/core/src/components/radio/radio.scss (hunk 1 above)
Comment (josokinas) [reply packages/core/src/components/radio/radio.scss:40]: I'm changing this to be with JSX.

File: packages/react/src/components/radio/radio.react.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        type="radio"
+        className={classy(c('radio-input'), m({ invalid }))}
+      />
+      <span aria-hidden="true">{/* custom "checked" indicator */}</span>
+      {(label || description) && (
+        <span>
+          <span>{label}</span>
+          <span>{description}</span>
+        </span>
+      )}
```
[the remark this answers is not in the document]
Comment (josokinas) [reply packages/react/src/components/radio/radio.react.tsx]: Changed to this/similar API now.

File: packages/core/src/helpers/border-radius/border-radius.scss [hunk 8]
```diff
@@ -1,8 +1,8 @@
-$allowed-sizes: ('small', 'medium', 'large', 'full');
+$_allowed-sizes: ('small', 'medium', 'large', 'full');
```
[the remark this answers is not in the document]
Comment (josokinas) [reply packages/core/src/helpers/border-radius/border-radius.scss:1]: Needed to... This is because now both `with-outer-focus` and `with-inner-focus` have same/similar variables, they cannot be imported together.

File: .stylelintrc.json [hunk 9]
```diff
@@ -3,8 +3,8 @@
   "rules": {
     "max-nesting-depth": [1, { "ignore": ["pseudo-classes"] }],
     "selector-class-pattern": "^-?[a-z]+(?:--?[a-z]+)*$",
-    "scss/dollar-variable-pattern": "^[a-z]+(?:-[a-z]+)*$",
-    "scss/at-function-pattern": "^[a-z]+(?:-[a-z]+)*$",
+    "scss/dollar-variable-pattern": "^_?[a-z]+(?:-[a-z]+)*$",
```
Comment (rabelloo) [inline .stylelintrc.json:6]: Maybe extract this (private scss variables) and the `space` change to another PR so we can quickly merge it (with nice commit messages)?

File: packages/core/src/components/radio/radio.scss [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+      --border-color: #{helpers.color('border-disabled')};
+      --color: #{helpers.color('content-disabled')};
+    }
+
+    > :not([aria-hidden='true']) {
+      padding-right: helpers.space(2) - $padding;
+    }
+  }
+
+  .ods-radio-input {
```
[the remark this answers is not in the document]
Comment (josokinas) [reply packages/core/src/components/radio/radio.scss:50]: Originally it was developed like such, however then there's issue with specificity, as you'd have e.g.:

```scss
.ods-radio {
  > input {
    &.-invalid { }
  }
}
```

Also, I feel like having two separate classes makes it more flexible.

File: packages/core/src/components/radio/radio.scss [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    --color: #{helpers.color('content-main')};
+
+    color: var(--color);
+    display: inline-flex;
+    gap: $padding;
+    padding: $padding;
+    position: relative;
+
+    &.-bordered {
+      --border-color: #{helpers.color('border-separator')};
```
[the remark this answers is not in the document]
Comment (josokinas) [reply packages/core/src/components/radio/radio.scss:24]: I did not, this is purely stylistic to align with other one.

Happy to revert if you object.

File: .stylelintrc.json (hunk 9 above)
Comment (josokinas) [reply .stylelintrc.json:6]: Sure:

- https://github.com/onfido/castor/pull/155
- https://github.com/onfido/castor/pull/156

File: packages/core/src/components/icon/icon.scss [hunk 12]
```diff
@@ -1,11 +1,9 @@
 @use '../../helpers';
+@use '../../mixins';
 
 @mixin Icon() {
-  $size: helpers.space(3);
-
   .ods-icon {
+    @include mixins.square(helpers.space(3));
```
[the remark this answers is not in the document]
Comment (josokinas) [reply packages/core/src/components/icon/icon.scss:6]: Not sure about this.

For example, it would not allow to use when size has to be calculated, or in case of something like `rem` or `em` should be used (which currently it would not support but still).

File: packages/core/src/components/radio/radio.scss [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+      --border-color: #{helpers.color('border-action-subtle')};
+    }
+
+    &.-disabled {
+      --background-color: unset;
+      --border-color: #{helpers.color('border-disabled')};
+      --color: #{helpers.color('content-disabled')};
+    }
+
+    > :not([aria-hidden='true']) {
```
[the remark this answers is not in the document]
Comment (josokinas) [reply packages/core/src/components/radio/radio.scss:45]: You're right, this indeed applies to `input` also...

I'm not a fan of using `*` as it's slow, however don't see other way.

One could be:

```scss
> span {
  padding-right: helpers.space(2) - $padding;
}
> [aria-hidden='true'] {
  padding: 0;
}
```

But also not great.

File: packages/core/src/components/icon/icon.scss (hunk 12 above)
[the remark this answers is not in the document]
Comment (rabelloo) [reply packages/core/src/components/icon/icon.scss:6]: Fair enough, let's leave as is for now.

File: packages/core/src/components/radio/radio.scss (hunk 11 above)
[the remark this answers is not in the document]
Comment (rabelloo) [reply packages/core/src/components/radio/radio.scss:24]: Which "other one" do you mean? `--color:`?

I actually had issues recently with something like `width: calc(100% - helper.space(4))` where it does need the `#{}` to properly interpolate.
So I don't mind that we use it everywhere for consistency, but I also don't mind being terser where possible.

Anyway, up to you.

File: packages/core/src/components/radio/radio.scss (hunk 13 above)
[the remark this answers is not in the document]
Comment (rabelloo) [reply packages/core/src/components/radio/radio.scss]: To be fair, CSS selector optimisation is misunderstood, and the universal rule `*` is one of the most prejudiced against.
According to multiple studies, the performance difference is negligible in most selectors.

https://meiert.com/en/blog/performance-of-css-selectors-2
> What we saw, namely, was an undue focus on selector performance and an outright childish ban of selectors, notably the universal selector.

https://www.stevesouders.com/blog/2009/03/10/performance-impact-of-css-selectors
> For most web sites, the possible performance gains from optimizing CSS selectors will be small, and are not worth the costs.

The focus should be on CSS rules that trigger repaint/reflow or that are just extremely unoptimised.

With that said, however, I don't like using `*` where it can be helped either.

I like your solution, overriding the value on more specific selectors.
However we'd need to include `input` too.

```scss
> input,
> [aria-hidden='true'] {
  padding: 0;
}
```

In which case the `:not` can work just as well
```scss
> :not(input),
> :not([aria-hidden='true']) {
  padding-right: helpers.space(2) - $padding;
}
```

**TL;DR:** I'm happy with the simplest approach which is not changing this selector, and including `padding: 0` on `.ods-radio-input`.

File: packages/core/src/components/radio/radio.scss (hunk 10 above)
[the remark this answers is not in the document]
Comment (rabelloo) [reply packages/core/src/components/radio/radio.scss:50]: Fair enough, specificity alone is a good reason.

File: packages/core/src/components/radio/radio.scss (hunk 11 above)
[the remark this answers is not in the document]
Comment (josokinas) [reply packages/core/src/components/radio/radio.scss:24]: Ah, sorry, I misread your comment.

It must be interpolated when using with CSS variables, however other places did not require such so I did remove.

File: packages/core/src/components/radio/radio.scss (hunk 13 above)
[the remark this answers is not in the document]
Comment (josokinas) [reply packages/core/src/components/radio/radio.scss]: Not a fan of `:not` here, would latest change (originally suggested by you) be fine? I've already pushed that change.

[the remark this answers is not in the document]
Comment (rabelloo) [reply packages/core/src/components/radio/radio.scss]: Yea that's fine. We might update it too if we go with the CSS `:checked` idea and it works.

File: packages/core/src/components/description/description.scss [hunk 14]
```diff
@@ -0,0 +1,14 @@
+@use '../../helpers';
+
+@mixin Description() {
+  .ods-description {
```
Comment (rabelloo) [inline packages/core/src/components/description/description.scss:4]: Are you still considering changing the name to `.ods-helper-text`?

File: packages/core/src/mixins/interactive.scss [hunk 15]
```diff
@@ -1,8 +1,11 @@
-@mixin interactive() {
-  cursor: pointer;
-  user-select: none;
+@mixin interactive($target: null) {
+  #{if($target, $target, '&')} {
+    cursor: pointer;
+    user-select: none;
+  }
```
Comment (rabelloo) [inline packages/core/src/mixins/interactive.scss:5]: Nice solution to reuse the mixin 👍🏻

**Review by rabelloo — APPROVED** [review]
Looks great!

File: packages/core/src/components/description/description.scss (hunk 14 above)
Comment (josokinas) [reply packages/core/src/components/description/description.scss:4]: I am yes, but will do as a separate PR.
